### PR TITLE
osd: update latest overlap to [0,obs.oi.size] before rollback_to #64735

### PR DESCRIPTION
**Pull request by yanggogo**
when rollback_to snap, we currently use the latest clone's current overlap to intersection_of older snapshot's clone overlap (which is not that correct). we should update the latest clone's current overlap to [0, obs.oi.size] before intersection_of other clone's overlap.

rbd du/rbd diff/rbd export-diff will be affected.

Fixes: https://tracker.ceph.com/issues/64735
Signed-off-by: <xingd@chinatelecom.cn>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -8501,16 +8501,26 @@ void PrimaryLogPG::_do_rollback_to(OpContext *ctx, ObjectContextRef rollback_to,
...
   t->add_obc(rollback_to);
 
+  if (obs.oi.size > 0) {
+    // change most recent clone_overlap to [0, obs.oi.size] before rollback to snap
+    if (ctx->new_snapset.clones.size() > 0) {
+      interval_set<uint64_t> rollback_overlaps;
+      rollback_overlaps.insert(0, obs.oi.size);
+      interval_set<uint64_t> &newest_overlap =
+        ctx->new_snapset.clone_overlap.rbegin()->second;
+      newest_overlap.swap(rollback_overlaps);
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:8511]: Modifying clone_overlaps directly here isn't appropriate -- that's what ctx->modified_ranges is for.  See the end of PrimaryLogPG::make_writeable.

Also, this logic looks wrong -- it looks like this last clone_overlap would indicate that the most recent clone and head *share* everything in the range [0, obs.oi-size).  Is that actually the intent?

In a larger sense, I'm not actually sure what about the logic below is wrong.  At a minimum, this PR needs a commit adding a unit test which fails with the current implementation.  It looks like you have a reproducer?

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -8501,16 +8501,26 @@ void PrimaryLogPG::_do_rollback_to(OpContext *ctx, ObjectContextRef rollback_to,
...
-	++iter)
-    overlaps.intersection_of(iter->second);
+	++iter) {
+        overlaps.intersection_of(iter->second);
+      }
 
-  if (obs.oi.size > 0) {
     interval_set<uint64_t> modified;
     modified.insert(0, obs.oi.size);
     overlaps.intersection_of(modified);
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:8526]: This looks right to me.
- overlaps is set of byte ranges that didn't change between the rollback target and head.
- intersection(overlaps, [0, obs.oi.size)) is the subset of the above which falls within the size of the current object
- [0, obs.oi.size) - intersection(overlaps, [0, obs.oi.size)) is therefore the set of ranges modified between the rollback target and head
- ctx->modified_ranges.union_of(modified); is the above plus any other extents modified in this transaction.

If you are seeing a bug with clone_overlaps, presumably one of the above steps is incorrect.

Comment (yanggogo) [reply src/osd/PrimaryLogPG.cc:8526]: yes, I found the latest overlap is not that correct when rollback_to snap.

```shell
[root@test xingd]# rbd create ceph-vm-pool-1/test_rollback -s 4M
[root@test xingd]# fio -ioengine=rbd -bs=4M -size=4M -rw=write -pool=ceph-vm-pool-1 -rbdname=test_rollback --group_report -name=test -direct=1 -numjobs=1 -iodepth=1
[root@test xingd]# rbd snap create ceph-vm-pool-1/test_rollback@base
[root@test xingd]# fio -ioengine=rbd -bs=2M -size=2M -rw=write -pool=ceph-vm-pool-1 -rbdname=test_rollback --group_report -name=test -direct=1 -numjobs=1 -iodepth=1
[root@test xingd]# rados -p ceph-vm-pool-1 listsnaps rbd_data.b4bf851b702e15.0000000000000000
rbd_data.b4bf851b702e15.0000000000000000:
cloneid	snaps	size	overlap
48	48	4194304	[2097152~2097152]
head	-	4194304
[root@test xingd]# rbd snap rollback ceph-vm-pool-1/test_rollback@base
[root@test xingd]# rados -p ceph-vm-pool-1 listsnaps rbd_data.b4bf851b702e15.0000000000000000
rbd_data.b4bf851b702e15.0000000000000000:
cloneid	snaps	size	overlap
48	48	4194304	[2097152~2097152]           ### after rollback， the overlap shoud be [0~4194304]
head	-	4194304
```

another test is:
```shell
[root@test xingd]# rbd create ceph-vm-pool-1/test_rollback1 -s 4M
[root@test xingd]# fio -ioengine=rbd -bs=4M -size=4M -rw=write -pool=ceph-vm-pool-1 -rbdname=test_rollback1 --group_report -name=test -direct=1 -numjobs=1 -iodepth=1
[root@test xingd]# rbd snap create ceph-vm-pool-1/test_rollback1@base
[root@test xingd]# rados -p ceph-vm-pool-1 listsnaps rbd_data.a9d86dcb31684.0000000000000000
rbd_data.a9d86dcb31684.0000000000000000:
cloneid	snaps	size	overlap
head	-	4194304
[root@test xingd]# rbd snap rollback ceph-vm-pool-1/test_rollback1@base
[root@test xingd]# rados -p ceph-vm-pool-1 listsnaps rbd_data.a9d86dcb31684.0000000000000000
rbd_data.a9d86dcb31684.0000000000000000:
cloneid	snaps	size	overlap
103	103	4194304	[0~4194304]                ### write nothing after snap create, and rollback to snap, the latest overlap is [0～4194304] （which is correct）
head	-	4194304
```

Comment (athanatos) [reply src/osd/PrimaryLogPG.cc:8526]: Ok, the first step would be to create a rados unit test in test/librados/snapshots_cxx.cc which demonstrates the problem.  Be sure to include cases where:
1. the rollback target is the newest clone
2. the rollback target is the most recent clone
3. the rollback target is the oldest clone
4. the rollback target is neither the oldest nor the newest clone

I think the actual bug is that the above loop should skip the last overlap entry.  The rollback overwrites the head, so we don't care about the overlap between head and the most recent snapshot -- we want the overlap between the most recent *clone* and the rollback target.  Once you have a unit test, try changing that.

Comment (athanatos) [reply src/osd/PrimaryLogPG.cc:8526]: > This looks right to me.
> 
>     * overlaps is set of byte ranges that didn't change between the rollback target and head.

In other words, this part is wrong.  We want overlap to be the set of byte ranges that didn't change between the rollback target and the *newest clone*, not head.

**Comment by Matan-B** [comment]
@athanatos, I'm not sure this is a bug, please see: https://tracker.ceph.com/issues/64735#note-3

**Review by Matan-B — CHANGES_REQUESTED** [review]
The change here seems to work only for a single clone case.
The following scenario seems to misbehave (w/ the suggesed change)
```
1) create 4M image
2) write 4M [A]
3) take snap_1
4) write 2M [B]
5) take snap_2
6) write 1M [C]

cloneid snaps   size    overlap
3       3       4194304 [2097152~2097152]
4       4       4194304 [1048576~3145728]
head    -       4194304
```

After rolling back to snap_1 we should expect the result to be:
```
cloneid snaps   size    overlap
3       3       4194304 [0~4194304]
4       4       4194304 [2097152~2097152]
head    -       4194304
```

However, the results is:
```
cloneid snaps   size    overlap
3       3       4194304 [2097152~2097152]
4       4       4194304 [2097152~2097152]
head    -       4194304
```

Am I missing something?

**Comment by Matan-B** [comment]
I think that the relevant bit that requires adjustment is in `PrimaryLogPG::make_writeable`:
```
 // update most recent clone_overlap and usage stats`.
 ...
 newest_overlap.subtract(ctx->modified_ranges);
```
Looking into this

**Comment by yanggogo** [comment]
> The change here seems to work only for a single clone case. The following scenario, when rollbacking to a snap other than the last one, seems to misbehave (w/ the suggesed change)
> 
> ```
> 1) create 4M image
> 2) write 4M [A]
> 3) take snap_1
> 4) write 2M [B]
> 5) take snap_2
> 6) write 1M [C]
> 
> cloneid snaps   size    overlap
> 3       3       4194304 [2097152~2097152]
> 4       4       4194304 [1048576~3145728]
> head    -       4194304
> ```
> 
> After rolling back to snap_1 we should expect the result to be:
> 
> ```
> cloneid snaps   size    overlap
> 3       3       4194304 [0~4194304]
> 4       4       4194304 [2097152~2097152]
> head    -       4194304
> ```
> 
> However, the results is:
> 
> ```
> cloneid snaps   size    overlap
> 3       3       4194304 [2097152~2097152]
> 4       4       4194304 [2097152~2097152]
> head    -       4194304
> ```
> 
> Am I missing something?

when rollback to snap, we use the older snaps's overlaps (intersec them, and oi.size - overlap = modified_ranges) to get the ***modified_ranges*** actually.
I think it's hard to get a perfect modifed_ranges easily when rollback to older snap, because rollback to is very fast, we don't ***write*** the HEAD actually.

That's what the unit test RollbackOverlap2PP try to explain.

1) write 32K to a rados object
2) take snap_1
3) write (offset=0,size=1k)         overlap [1k~31k]
4) take snap_2
5) write (offset=31k,size=1k)       overlap [0~31k]
6) take snap_3
7) write (offset=16k,size=1k)       overlap [0~16k,17k~15k]
8) take snap_4
9)           rollback or write                       the latest overlap [xxx]

we never change the older snap's overlap when rollback to snap, we only update the ***latest*** snap's overlap.

when rollback to snap_4      the latest overlap is [0,32k]
when rollback to snap_3      the latest overlap is [0~16k,17k~15k]
when rollback to snap_2      the latest overlap is intersec([0~31k], [0~16k, 17k~15k]) = [0~16k, 17k~14k]
when rollback to snap_1      the latest overlap is intersec([1k~31k], [0~31k], [0~16k, 17k~15k]) = ([1k~15k,17k~14k])


> The change here seems to work only for a single clone case. The following scenario seems to misbehave (w/ the suggesed change)
> 
> ```
> 1) create 4M image
> 2) write 4M [A]
> 3) take snap_1
> 4) write 2M [B]
> 5) take snap_2
> 6) write 1M [C]
> 
> cloneid snaps   size    overlap
> 3       3       4194304 [2097152~2097152]
> 4       4       4194304 [1048576~3145728]
> head    -       4194304
> ```
> 
> After rolling back to snap_1 we should expect the result to be:
> 
> ```
> cloneid snaps   size    overlap
> 3       3       4194304 [0~4194304]
> 4       4       4194304 [2097152~2097152]
> head    -       4194304
> ```
> 
> However, the results is:
> 
> ```
> cloneid snaps   size    overlap
> 3       3       4194304 [2097152~2097152]
> 4       4       4194304 [2097152~2097152]
> head    -       4194304
> ```
> 
> Am I missing something?

when rollback to snap, we use the older snaps's overlaps (intersec them, and oi.size - overlap = modified_ranges) to get the ***modified_ranges*** actually.
I think it's hard to get a perfect modifed_ranges easily when rollback to older snap, because rollback to is very fast, we don't ***write*** the HEAD actually.

That's what the unit test RollbackOverlap2PP try to explain.

1) write 32K to a rados object
2) take snap_1
3) write (offset=0,size=1k)         overlap [1k~31k]
4) take snap_2
5) write (offset=31k,size=1k)       overlap [0~31k]
6) take snap_3
7) write (offset=16k,size=1k)       overlap [0~16k,17k~15k]
8) take snap_4
9)  rollback or write something          the latest overlap [xxx]

we never change the older snap's overlap when rollback to snap, we only update the ***latest*** snap's overlap.

when rollback to snap_4      the latest overlap is [0,32k]
when rollback to snap_3      the latest overlap is [0~16k,17k~15k]
when rollback to snap_2      the latest overlap is intersec([0~31k], [0~16k, 17k~15k]) = [0~16k, 17k~14k]
when rollback to snap_1      the latest overlap is intersec([1k~31k], [0~31k], [0~16k, 17k~15k]) = ([1k~15k,17k~14k])


> The change here seems to work only for a single clone case. The following scenario, when rollbacking to a snap other than the last one, seems to misbehave (w/ the suggesed change)
> 
> ```
> 1) create 4M image
> 2) write 4M [A]
> 3) take snap_1
> 4) write 2M [B]
> 5) take snap_2
> 6) write 1M [C]
> 
> cloneid snaps   size    overlap
> 3       3       4194304 [2097152~2097152]
> 4       4       4194304 [1048576~3145728]
> head    -       4194304
> ```
> 
> After rolling back to snap_1 we should expect the result to be:
> 
> ```
> cloneid snaps   size    overlap
> 3       3       4194304 [0~4194304]
> 4       4       4194304 [2097152~2097152]
> head    -       4194304
> ```
> 
> However, the results is:
> 
> ```
> cloneid snaps   size    overlap
> 3       3       4194304 [2097152~2097152]
> 4       4       4194304 [2097152~2097152]
> head    -       4194304
> ```
> 
> Am I missing something?

when rollback to snap, we use the older snaps's overlaps (intersec them, and oi.size - overlap = modified_ranges) to get the ***modified_ranges*** actually.
I think it's hard to get a perfect modifed_ranges when rollback to older snap, because rollback to is very fast, we don't ***write*** the HEAD actually.

That's what the unit test RollbackOverlap2PP try to explain.

1) write 32K to a rados object
2) take snap_1
3) write (offset=0,size=1k)         overlap [1k~31k]
4) take snap_2
5) write (offset=31k,size=1k)       overlap [0~31k]
6) take snap_3
7) write (offset=16k,size=1k)       overlap [0~16k,17k~15k]
8) take snap_4
9)  rollback or write something          the latest overlap [xxx]

we never change the older snap's overlap when rollback to snap, we only update the ***latest*** snap's overlap.

when rollback to snap_4      the latest overlap is [0,32k]
when rollback to snap_3      the latest overlap is [0~16k,17k~15k]
when rollback to snap_2      the latest overlap is intersec([0~31k], [0~16k, 17k~15k]) = [0~16k, 17k~14k]
when rollback to snap_1      the latest overlap is intersec([1k~31k], [0~31k], [0~16k, 17k~15k]) = ([1k~15k,17k~14k])

**Comment by Matan-B** [comment]
> we never change the older snap's overlap when rollback to snap, we only update the _**latest**_ snap's overlap.

I'm not sure this is the correct way, we will probably have to revisit the clone_overlap handling on _any_ rollback case.

**Comment by athanatos** [comment]
My original comment hasn't been addressed.  The unit tests are a bit tough to read -- I'm going to create a branch refining this PR a bit in the next few days.  I'm marking it DNM in the mean time.

**Comment by yanggogo** [comment]
> My original comment hasn't been addressed. The unit tests are a bit tough to read -- I'm going to create a branch refining this PR a bit in the next few days. I'm marking it DNM in the mean time.

OK, Thanks. "I think the actual bug is that the above loop should skip the last overlap entry" --- this is an another way to fix it.

**Comment by Matan-B** [comment]
Closing for #56696

**Pull request closed without merging**